// @intellectronica/ruler/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SKILLZ_MCP_SERVER_NAME = exports.SKILL_MD_FILENAME = exports.SKILLZ_DIR = exports.CLAUDE_SKILLS_PATH = exports.RULER_SKILLS_PATH = exports.SKILLS_DIR = exports.DEFAULT_RULES_FILENAME = exports.ERROR_PREFIX = void 0;
exports.actionPrefix = actionPrefix;
exports.createRulerError = createRulerError;
exports.logVerbose = logVerbose;
exports.logInfo = logInfo;
exports.logWarn = logWarn;
exports.logError = logError;
exports.logVerboseInfo = logVerboseInfo;
exports.ERROR_PREFIX = '[ruler]';
// Centralized default rules filename. Now points to 'AGENTS.md'.
// Legacy '.ruler/instructions.md' is still supported as a fallback with a warning.
exports.DEFAULT_RULES_FILENAME = 'AGENTS.md';
function actionPrefix(dry) {
    return dry ? '[ruler:dry-run]' : '[ruler]';
}
function createRulerError(message, context) {
    const fullMessage = context
        ? `${exports.ERROR_PREFIX} ${message} (Context: ${context})`
        : `${exports.ERROR_PREFIX} ${message}`;
    return new Error(fullMessage);
}
function logVerbose(message, isVerbose) {
    if (isVerbose) {
        console.error(`[ruler:verbose] ${message}`);
    }
}
/**
 * Centralized logging functions with consistent output streams and prefixing.
 * - info/verbose go to stdout (user-visible progress)
 * - warn/error go to stderr (problems)
 */
function logInfo(message, dryRun = false) {
    const prefix = actionPrefix(dryRun);
    console.log(`${prefix} ${message}`);
}
function logWarn(message, dryRun = false) {
    const prefix = actionPrefix(dryRun);
    console.warn(`${prefix} ${message}`);
}
function logError(message, dryRun = false) {
    const prefix = actionPrefix(dryRun);
    console.error(`${prefix} ${message}`);
}
function logVerboseInfo(message, isVerbose, dryRun = false) {
    if (isVerbose) {
        const prefix = actionPrefix(dryRun);
        console.log(`${prefix} ${message}`);
    }
}
// Skills-related constants
exports.SKILLS_DIR = 'skills';
exports.RULER_SKILLS_PATH = '.ruler/skills';
exports.CLAUDE_SKILLS_PATH = '.claude/skills';
exports.SKILLZ_DIR = '.skillz';
exports.SKILL_MD_FILENAME = 'SKILL.md';
exports.SKILLZ_MCP_SERVER_NAME = 'skillz';
