// cursor-background-agent-mcp-server@1.0.3/dist/domain/manifest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.manifest = void 0;
exports.manifest = {
    tools: [
        {
            name: "launchAgent",
            description: "Start a new background agent to work on a repository. IMPORTANT: The repository parameter must be a full GitHub URL (e.g., 'https://github.com/owner/repo') and you must have access to it with the Cursor GitHub App installed.",
            inputSchema: {
                type: "object",
                properties: {
                    prompt: {
                        type: "object",
                        properties: {
                            text: { type: "string" },
                            images: {
                                type: "array",
                                items: {
                                    type: "object",
                                    properties: {
                                        data: { type: "string" },
                                        dimension: {
                                            type: "object",
                                            properties: {
                                                width: { type: "number" },
                                                height: { type: "number" }
                                            },
                                            required: ["width", "height"]
                                        }
                                    },
                                    required: ["data", "dimension"]
                                }
                            }
                        },
                        required: ["text"]
                    },
                    source: {
                        type: "object",
                        properties: {
                            repository: {
                                type: "string",
                                description: "Full GitHub repository URL in format: https://github.com/owner/repo-name. You must have access to this repository with the Cursor GitHub App installed.",
                                pattern: "^https://github\\.com/[^/]+/[^/]+$"
                            },
                            ref: {
                                type: "string",
                                description: "Git branch or commit reference (usually 'main' or 'master')"
                            }
                        },
                        required: ["repository", "ref"]
                    }
                },
                required: ["prompt", "source"]
            },
            outputSchema: {
                type: "object",
                properties: {
                    id: { type: "string" },
                    name: { type: "string" },
                    status: { type: "string" },
                    source: { type: "object" },
                    target: { type: "object" },
                    createdAt: { type: "string" }
                },
                required: ["id", "name", "status", "source", "target", "createdAt"]
            },
            exampleInput: {
                prompt: {
                    text: "Add a README.md file with installation instructions"
                },
                source: {
                    repository: "https://github.com/microsoft/vscode",
                    ref: "main"
                }
            },
            exampleOutput: {
                id: "bc-1eb2cd96-aa10-46a1-a4d9-dc5c911374f6",
                name: "Add README Documentation",
                status: "CREATING",
                source: { repository: "https://github.com/microsoft/vscode", ref: "main" },
                target: { branchName: "cursor/add-readme-1234", url: "https://cursor.com/agents?id=bc-1eb2cd96-aa10-46a1-a4d9-dc5c911374f6", autoCreatePr: false },
                createdAt: "2024-01-15T10:30:00Z"
            }
        },
        {
            name: "listAgents",
            description: "Retrieve a paginated list of all background agents for the authenticated user.",
            inputSchema: { type: "object", properties: {}, additionalProperties: false },
            outputSchema: {
                type: "object",
                properties: {
                    agents: { type: "array", items: { type: "object" } },
                    nextCursor: { type: "string" }
                },
                required: ["agents"]
            }
        },
        {
            name: "listModels",
            description: "Retrieve a list of recommended AI models that can be used for background agents. Note: This requires valid authentication.",
            inputSchema: { type: "object", properties: {}, additionalProperties: false },
            outputSchema: {
                type: "object",
                properties: {
                    models: { type: "array", items: { type: "object" } }
                },
                required: ["models"]
            }
        },
        {
            name: "addFollowup",
            description: "Send an additional instruction to a running background agent.",
            inputSchema: {
                type: "object",
                properties: {
                    agentId: { type: "string" },
                    prompt: {
                        type: "object",
                        properties: {
                            text: { type: "string" },
                            images: {
                                type: "array",
                                items: {
                                    type: "object",
                                    properties: {
                                        data: { type: "string" },
                                        dimension: {
                                            type: "object",
                                            properties: {
                                                width: { type: "number" },
                                                height: { type: "number" }
                                            },
                                            required: ["width", "height"]
                                        }
                                    },
                                    required: ["data", "dimension"]
                                }
                            }
                        },
                        required: ["text"]
                    }
                },
                required: ["agentId", "prompt"]
            },
            outputSchema: {
                type: "object",
                properties: {
                    success: { type: "boolean" }
                },
                required: ["success"]
            }
        },
        {
            name: "getAgentConversation",
            description: "Retrieve the conversation history of a background agent.",
            inputSchema: {
                type: "object",
                properties: {
                    agentId: { type: "string" }
                },
                required: ["agentId"]
            },
            outputSchema: {
                type: "object",
                properties: {
                    id: { type: "string" },
                    messages: { type: "array", items: { type: "object" } }
                },
                required: ["id", "messages"]
            }
        },
        {
            name: "getAgentStatus",
            description: "Get the current status and results of a specific background agent.",
            inputSchema: {
                type: "object",
                properties: {
                    agentId: { type: "string" }
                },
                required: ["agentId"]
            },
            outputSchema: {
                type: "object",
                properties: {
                    id: { type: "string" },
                    name: { type: "string" },
                    status: { type: "string" },
                    source: { type: "object" },
                    target: { type: "object" },
                    summary: { type: "string" },
                    createdAt: { type: "string" }
                },
                required: ["id", "name", "status", "source", "target", "createdAt"]
            }
        },
        {
            name: "listRepositories",
            description: "Retrieve a list of GitHub repositories accessible to the authenticated user. This endpoint is rate-limited: 1/minute, 30/hour per user. May take tens of seconds to respond for users with many repos. Handles unavailability gracefully.",
            inputSchema: { type: "object", properties: {}, additionalProperties: false },
            outputSchema: {
                type: "object",
                properties: {
                    repositories: { type: "array", items: { type: "object" } },
                    error: { type: "string" }
                }
            }
        }
    ],
    resources: []
};
