// @intellectronica/ruler@0.3.16/dist/core/SkillsProcessor.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.discoverSkills = discoverSkills;
exports.getSkillsGitignorePaths = getSkillsGitignorePaths;
exports.propagateSkills = propagateSkills;
exports.propagateSkillsForClaude = propagateSkillsForClaude;
exports.propagateSkillsForSkillz = propagateSkillsForSkillz;
exports.buildSkillzMcpConfig = buildSkillzMcpConfig;
const path = __importStar(require("path"));
const fs = __importStar(require("fs/promises"));
const constants_1 = require("../constants");
const SkillsUtils_1 = require("./SkillsUtils");
/**
 * Discovers skills in the project's .ruler/skills directory.
 * Returns discovered skills and any validation warnings.
 */
async function discoverSkills(projectRoot) {
    const skillsDir = path.join(projectRoot, constants_1.RULER_SKILLS_PATH);
    // Check if skills directory exists
    try {
        await fs.access(skillsDir);
    }
    catch {
        // Skills directory doesn't exist - this is fine, just return empty
        return { skills: [], warnings: [] };
    }
    // Walk the skills tree
    return await (0, SkillsUtils_1.walkSkillsTree)(skillsDir);
}
/**
 * Gets the paths that skills will generate, for gitignore purposes.
 * Returns empty array if skills directory doesn't exist.
 */
async function getSkillsGitignorePaths(projectRoot) {
    const skillsDir = path.join(projectRoot, constants_1.RULER_SKILLS_PATH);
    // Check if skills directory exists
    try {
        await fs.access(skillsDir);
    }
    catch {
        return [];
    }
    // Import here to avoid circular dependency
    const { CLAUDE_SKILLS_PATH, SKILLZ_DIR } = await Promise.resolve().then(() => __importStar(require('../constants')));
    return [
        path.join(projectRoot, CLAUDE_SKILLS_PATH),
        path.join(projectRoot, SKILLZ_DIR),
    ];
}
/**
 * Module-level state to track if experimental warning has been shown.
 * This ensures the warning appears once per process (CLI invocation), not once per apply call.
 * This is intentional: warnings about experimental features should not spam the user
 * if they run multiple applies in the same process or test suite.
 */
let hasWarnedExperimental = false;
/**
 * Warns once per process about experimental skills features and uv requirement.
 * Uses module-level state to prevent duplicate warnings within the same process.
 */
function warnOnceExperimentalAndUv(verbose, dryRun) {
    if (hasWarnedExperimental) {
        return;
    }
    hasWarnedExperimental = true;
    (0, constants_1.logWarn)('Skills support is experimental and behavior may change in future releases.', dryRun);
    (0, constants_1.logWarn)('Skills MCP server (Skillz) requires uv. Install: https://github.com/astral-sh/uv', dryRun);
}
/**
 * Cleans up skills directories (.claude/skills and .skillz) when skills are disabled.
 * This ensures that stale skills from previous runs don't persist when skills are turned off.
 */
async function cleanupSkillsDirectories(projectRoot, dryRun, verbose) {
    const claudeSkillsPath = path.join(projectRoot, constants_1.CLAUDE_SKILLS_PATH);
    const skillzPath = path.join(projectRoot, constants_1.SKILLZ_DIR);
    // Clean up .claude/skills
    try {
        await fs.access(claudeSkillsPath);
        if (dryRun) {
            (0, constants_1.logVerboseInfo)(`DRY RUN: Would remove ${constants_1.CLAUDE_SKILLS_PATH}`, verbose, dryRun);
        }
        else {
            await fs.rm(claudeSkillsPath, { recursive: true, force: true });
            (0, constants_1.logVerboseInfo)(`Removed ${constants_1.CLAUDE_SKILLS_PATH} (skills disabled)`, verbose, dryRun);
        }
    }
    catch {
        // Directory doesn't exist, nothing to clean
    }
    // Clean up .skillz
    try {
        await fs.access(skillzPath);
        if (dryRun) {
            (0, constants_1.logVerboseInfo)(`DRY RUN: Would remove ${constants_1.SKILLZ_DIR}`, verbose, dryRun);
        }
        else {
            await fs.rm(skillzPath, { recursive: true, force: true });
            (0, constants_1.logVerboseInfo)(`Removed ${constants_1.SKILLZ_DIR} (skills disabled)`, verbose, dryRun);
        }
    }
    catch {
        // Directory doesn't exist, nothing to clean
    }
}
/**
 * Propagates skills for agents that need them.
 */
async function propagateSkills(projectRoot, agents, skillsEnabled, verbose, dryRun) {
    if (!skillsEnabled) {
        (0, constants_1.logVerboseInfo)('Skills support disabled, cleaning up skills directories', verbose, dryRun);
        // Clean up skills directories when skills are disabled
        await cleanupSkillsDirectories(projectRoot, dryRun, verbose);
        return;
    }
    const skillsDir = path.join(projectRoot, constants_1.RULER_SKILLS_PATH);
    // Check if skills directory exists
    try {
        await fs.access(skillsDir);
    }
    catch {
        // No skills directory - this is fine
        (0, constants_1.logVerboseInfo)('No .ruler/skills directory found, skipping skills propagation', verbose, dryRun);
        return;
    }
    // Discover skills
    const { skills, warnings } = await discoverSkills(projectRoot);
    if (warnings.length > 0) {
        warnings.forEach((warning) => (0, constants_1.logWarn)(warning, dryRun));
    }
    if (skills.length === 0) {
        (0, constants_1.logVerboseInfo)('No valid skills found in .ruler/skills', verbose, dryRun);
        return;
    }
    (0, constants_1.logVerboseInfo)(`Discovered ${skills.length} skill(s)`, verbose, dryRun);
    // Check if any agents need skills
    const hasNativeSkillsAgent = agents.some((a) => a.supportsNativeSkills?.());
    const hasMcpAgent = agents.some((a) => a.supportsMcpStdio?.() && !a.supportsNativeSkills?.());
    if (!hasNativeSkillsAgent && !hasMcpAgent) {
        (0, constants_1.logVerboseInfo)('No agents require skills support', verbose, dryRun);
        return;
    }
    // Warn about experimental features
    if (hasMcpAgent) {
        warnOnceExperimentalAndUv(verbose, dryRun);
    }
    // Copy to Claude skills directory if needed
    if (hasNativeSkillsAgent) {
        (0, constants_1.logVerboseInfo)(`Copying skills to ${constants_1.CLAUDE_SKILLS_PATH} for Claude Code`, verbose, dryRun);
        await propagateSkillsForClaude(projectRoot, { dryRun });
    }
    // Copy to .skillz directory if needed
    if (hasMcpAgent) {
        (0, constants_1.logVerboseInfo)(`Copying skills to ${constants_1.SKILLZ_DIR} for MCP agents`, verbose, dryRun);
        await propagateSkillsForSkillz(projectRoot, { dryRun });
    }
}
/**
 * Propagates skills for Claude Code by copying .ruler/skills to .claude/skills.
 * Uses atomic replace to ensure safe overwriting of existing skills.
 * Returns dry-run steps if dryRun is true, otherwise returns empty array.
 */
async function propagateSkillsForClaude(projectRoot, options) {
    const skillsDir = path.join(projectRoot, constants_1.RULER_SKILLS_PATH);
    const claudeSkillsPath = path.join(projectRoot, constants_1.CLAUDE_SKILLS_PATH);
    const claudeDir = path.dirname(claudeSkillsPath);
    // Check if source skills directory exists
    try {
        await fs.access(skillsDir);
    }
    catch {
        // No skills directory - return empty
        return [];
    }
    if (options.dryRun) {
        return [`Copy skills from ${constants_1.RULER_SKILLS_PATH} to ${constants_1.CLAUDE_SKILLS_PATH}`];
    }
    // Ensure .claude directory exists
    await fs.mkdir(claudeDir, { recursive: true });
    // Use atomic replace: copy to temp, then rename
    const tempDir = path.join(claudeDir, `skills.tmp-${Date.now()}`);
    try {
        // Copy to temp directory
        await (0, SkillsUtils_1.copySkillsDirectory)(skillsDir, tempDir);
        // Atomically replace the target
        // First, remove existing target if it exists
        try {
            await fs.rm(claudeSkillsPath, { recursive: true, force: true });
        }
        catch {
            // Target didn't exist, that's fine
        }
        // Rename temp to target
        await fs.rename(tempDir, claudeSkillsPath);
    }
    catch (error) {
        // Clean up temp directory on error
        try {
            await fs.rm(tempDir, { recursive: true, force: true });
        }
        catch {
            // Ignore cleanup errors
        }
        throw error;
    }
    return [];
}
/**
 * Propagates skills for MCP agents by copying .ruler/skills to .skillz.
 * Uses atomic replace to ensure safe overwriting of existing skills.
 * Returns dry-run steps if dryRun is true, otherwise returns empty array.
 */
async function propagateSkillsForSkillz(projectRoot, options) {
    const skillsDir = path.join(projectRoot, constants_1.RULER_SKILLS_PATH);
    const skillzPath = path.join(projectRoot, constants_1.SKILLZ_DIR);
    // Check if source skills directory exists
    try {
        await fs.access(skillsDir);
    }
    catch {
        // No skills directory - return empty
        return [];
    }
    if (options.dryRun) {
        return [
            `Copy skills from ${constants_1.RULER_SKILLS_PATH} to ${constants_1.SKILLZ_DIR}`,
            `Configure Skillz MCP server with absolute path to ${constants_1.SKILLZ_DIR}`,
        ];
    }
    // Use atomic replace: copy to temp, then rename
    const tempDir = path.join(projectRoot, `${constants_1.SKILLZ_DIR}.tmp-${Date.now()}`);
    try {
        // Copy to temp directory
        await (0, SkillsUtils_1.copySkillsDirectory)(skillsDir, tempDir);
        // Atomically replace the target
        // First, remove existing target if it exists
        try {
            await fs.rm(skillzPath, { recursive: true, force: true });
        }
        catch {
            // Target didn't exist, that's fine
        }
        // Rename temp to target
        await fs.rename(tempDir, skillzPath);
    }
    catch (error) {
        // Clean up temp directory on error
        try {
            await fs.rm(tempDir, { recursive: true, force: true });
        }
        catch {
            // Ignore cleanup errors
        }
        throw error;
    }
    return [];
}
/**
 * Builds MCP config for Skillz server.
 */
function buildSkillzMcpConfig(projectRoot) {
    const skillzAbsPath = path.resolve(projectRoot, constants_1.SKILLZ_DIR);
    return {
        [constants_1.SKILLZ_MCP_SERVER_NAME]: {
            command: 'uvx',
            args: ['skillz@latest', skillzAbsPath],
        },
    };
}
